Shared Filesystems: Share Export Locations

pospispa opened the pull request:
For #380

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

https://github.com/openstack/manila/blob/master/manila/api/v2/share_export_locations.py#L25

jtopjian commented on openstack/sharedfilesystems/v2/shares/results.go:
```diff
@@ -110,3 +110,31 @@ type DeleteResult struct {
+	Path string `json:"path"`
+	// The UUID of the share instance that this export location belongs to.
+	ShareInstanceID string `json:"share_instance_id"`
+	// Defines purpose of an export location. If set to true, then it is expected to be used for service needs and by administrators only. If it is set to false, then this export location can be used by end users.
```
nit: Can you wrap this comment to around 80 characters?

jtopjian commented on openstack/sharedfilesystems/v2/shares/results.go:
```diff
@@ -110,3 +110,31 @@ type DeleteResult struct {
+	IsAdminOnly bool `json:"is_admin_only"`
+	// The share export location UUID.
+	ID string `json:"id"`
+	// Drivers may use this field to identify which export locations are most efficient and should be used preferentially by clients. By default it is set to false value. New in version 2.14
```
nit: Can you wrap this comment to around 80 characters?

jtopjian commented on openstack/sharedfilesystems/v2/shares/results.go:
```diff
@@ -110,3 +110,31 @@ type DeleteResult struct {
+// ExtractExportLocations will get the Export Locations from the commonResult
+func (r commonResult) ExtractExportLocations() ([]ExportLocation, error) {
+	var s struct {
+		GetExportLocationsRes []ExportLocation `json:"export_locations"`
```
I think this should be:

```go
ExportLocations  []ExportLocation `json:"export_locations"`
```

pospispa commented:
@jtopjian I believe I have addressed all your comments and I created https://github.com/gophercloud/gophercloud/issues/447 in order to add a microversion check later.

jtopjian commented:
This looks good to me. I was able to get a share created in a test environment and the API results matched with what's here and in the referenced Manila code.

jrperritt commented on openstack/sharedfilesystems/v2/shares/results.go:
```diff
@@ -110,3 +110,36 @@ type DeleteResult struct {
+}
+
+// ExtractExportLocations will get the Export Locations from the commonResult
+func (r commonResult) ExtractExportLocations() ([]ExportLocation, error) {
```
Why is this a `commonResult` instead of a `GetExportLocationResult`? Can every response type extract this data?

jtopjian replied:
I caught that as well last night and I saw something else in the sharedfilesystems packages that made me think it was ok, but I most likely got it mixed up with the same occurrence in #346.

I doubt this is function is applicable to all. I don't see how a `share` result can extract this information. I can follow-up with a PR to fix.

jtopjian replied:
Ah! This is what it was: https://github.com/gophercloud/gophercloud/blob/master/openstack/sharedfilesystems/v2/sharenetworks/results.go#L172-L182

But upon reading further, I see that those actions return a share network and not some other result. So this is my mistake for not reading further.
